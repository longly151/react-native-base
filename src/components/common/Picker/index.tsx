import React, { useMemo, useState, useRef, useImperativeHandle, useEffect } from 'react';
import { ColorValue, Platform, View, TouchableOpacity, StyleProp, TextStyle } from 'react-native';
import withTailwind from '@components/hoc/withTailwind';
import { useTranslation } from 'react-i18next';
import { AdditionalTextInputProps, handleInputStyle, ReactHookFormProps } from '../Input';
import { useTheme } from '@react-navigation/native';
import Text from '../Text';
import Icon from '../Icon';
import { ActionSheetOptions, useActionSheet } from '@expo/react-native-action-sheet';
import { useForm, useController } from 'react-hook-form';
import Helper from '@utils/Helper';
import ModalPicker, { ModalPickerProps } from './ModalPicker';
import { ModalType } from '../Modal';
import _ from 'lodash';

export interface BasePickerProps
  extends ActionSheetOptions,
    AdditionalTextInputProps,
    ReactHookFormProps {
  className?: string;
  editable?: boolean;
  placeholder?: string;
  placeholderTextColor?: ColorValue;
  onCancel?: () => void;
  style?: StyleProp<TextStyle> | undefined;
  useModal?: boolean;
  modalPickerProps?: ModalPickerProps;
}

interface SingleSelect {
  type: 'single';
  selectedIndex?: number;
  onConfirm?: (buttonIndex: number) => void;
  convertValueToText?: (buttonIndex: number) => string;
}

interface MultipleSelect {
  type: 'multiple';
  selectedIndexes?: number[];
  onConfirm?: (buttonIndexes: number[]) => void;
  convertValueToText?: (buttonIndexes: number[]) => string;
}

export type PickerProps = BasePickerProps & (SingleSelect | MultipleSelect);

const Picker: React.FC<PickerProps> = React.forwardRef((props, ref: any) => {
  let pickerRef = useRef<View>(null);
  const modalPickerRef = useRef<ModalType>(null);

  const theme = useTheme();
  const { t } = useTranslation();
  const { showActionSheetWithOptions } = useActionSheet();

  const {
    title,
    titleProps,
    errorProps,
    leftIconProps,
    rightIconProps,
    name,
    type,
    // @ts-ignore
    selectedIndex: selectedIndexProp,
    // @ts-ignore
    selectedIndexes: selectedIndexesProp,
    convertValueToText: convertValueToTextProp,
    editable,
    onConfirm,
    onCancel,
    placeholder,
    options: optionsProp,
    control,
    nextField,
    rules,
    useModal,
    modalPickerProps,
    ...otherProps
  } = props;

  const handledStyle = useMemo(() => handleInputStyle(props, theme), [props, theme]);
  const { placeholderTextColor, inactiveBorderColor, activeBorderColor } = handledStyle;

  const [selectedIndex, setSelectedIndex] = useState<number>(selectedIndexProp);
  const [selectedIndexes, setSelectedIndexes] = useState<number[]>(selectedIndexesProp || []);

  const [borderColor, setBorderColor] = useState(
    !props.autoFocus ? inactiveBorderColor : activeBorderColor,
  );
  useEffect(() => {
    setBorderColor(inactiveBorderColor);
  }, [inactiveBorderColor]);

  const { control: formControl } = useForm({ mode: 'onTouched' });
  const { field, fieldState, formState } = useController({
    control: control || formControl,
    defaultValue: selectedIndexProp,
    rules,
    name: name || '',
  });

  const onOpenActionSheet = async () => {
    if (useModal || type === 'multiple') {
      modalPickerRef.current?.trigger(true);
      return;
    }

    if (!fieldState.error) {
      setBorderColor(activeBorderColor);
      await Helper.sleep(100);
    }
    showActionSheetWithOptions(
      {
        options: [...optionsProp, t('common.cancel')],
        destructiveButtonIndex: Platform.OS === 'android' ? optionsProp.length : undefined,
        cancelButtonIndex: optionsProp.length,
        userInterfaceStyle: theme.dark ? 'dark' : 'light',
        containerStyle: { backgroundColor: theme.colors.card },
        textStyle: { color: theme.colors.text },
        title,
      },
      async buttonIndex => {
        if (buttonIndex !== undefined && buttonIndex !== optionsProp.length) {
          if (name && control) {
            field.onChange(buttonIndex);
            field.onBlur();

            setSelectedIndex(buttonIndex);
            onConfirm?.(buttonIndex as any);

            const isValidated = await nextField?.trigger(name);

            if (isValidated) {
              setBorderColor(inactiveBorderColor);
              if (nextField && !formState.touchedFields[nextField.name]) {
                setTimeout(
                  () => {
                    nextField?.setFocus(nextField.name);
                  },
                  Platform.OS === 'android' ? 0 : 300,
                );
              }
            }
          } else {
            setSelectedIndex(buttonIndex);
            onConfirm?.(buttonIndex as any);
          }
        } else {
          const isValidated = await nextField?.trigger(name);
          if (isValidated) {
            setBorderColor(inactiveBorderColor);
          }
          if (name && control) {
            field.onBlur();
          }
          onCancel?.();
        }
      },
    );
  };

  const convertValueToText = () => {
    let text = placeholder;

    if (selectedIndex !== undefined) {
      if (convertValueToTextProp) {
        text = convertValueToTextProp(selectedIndex as any);
      } else {
        text = optionsProp[selectedIndex];
      }
    } else if (!_.isEmpty(selectedIndexes)) {
      if (convertValueToTextProp) {
        text = convertValueToTextProp(selectedIndexes as any);
      } else {
        text = optionsProp.filter((__, index) => selectedIndexes.includes(index)).join(', ');
      }
    }
    return text;
  };

  useImperativeHandle(ref, () => ({
    ...pickerRef.current,
    selectedIndex,
    setSelectedIndex,
    focus: onOpenActionSheet,
  }));

  useEffect(() => {
    if (name && control && fieldState.error && borderColor !== theme.colors.error) {
      setBorderColor(theme.colors.error);
    }
  }, [borderColor, control, fieldState.error, name, theme.colors]);

  const renderErrors = () => {
    if (name && control && fieldState.error) {
      return (
        <Text
          testID="ErrorText"
          color={theme.colors.error}
          {...errorProps}
          style={handledStyle.errorStyles}
        >
          {fieldState.error.message}
        </Text>
      );
    }
    return null;
  };

  const handleRef = (e: any) => {
    if (e) {
      e.focus = onOpenActionSheet;
    }
    field.ref(e);
    pickerRef = {
      current: e,
    };
  };

  return (
    <View testID="ContainerView" ref={handleRef} style={handledStyle.containerStyle}>
      {title ? (
        <Text testID="TitleText" {...titleProps} style={handledStyle.titleStyles}>
          {title}
        </Text>
      ) : null}

      <TouchableOpacity
        testID="PickerButton"
        activeOpacity={0.6}
        style={[handledStyle.inputContainerStyle, { borderColor }]}
        onPress={onOpenActionSheet}
        disabled={!editable}
      >
        {leftIconProps ? (
          <Icon
            testID="LeftIcon"
            color={handledStyle.otherStyle.color}
            size={handledStyle.otherStyle.fontSize}
            {...leftIconProps}
            style={[handledStyle.leftIconStyles, leftIconProps.style]}
          />
        ) : null}
        <Text
          testID="OptionText"
          {...(otherProps as any)}
          style={[
            handledStyle.otherStyle,
            {
              color:
                selectedIndex !== undefined || !_.isEmpty(selectedIndexes)
                  ? handledStyle.otherStyle.color
                  : placeholderTextColor,
            },
          ]}
        >
          {convertValueToText()}
        </Text>
        {rightIconProps ? (
          <Icon
            testID="RightIcon"
            color={handledStyle.otherStyle.color}
            size={handledStyle.otherStyle.fontSize}
            {...rightIconProps}
            style={[handledStyle.rightIconStyles, rightIconProps.style]}
          />
        ) : null}
      </TouchableOpacity>
      {renderErrors()}
      <ModalPicker
        ref={modalPickerRef}
        {...modalPickerProps}
        options={optionsProp}
        selectedIndex={selectedIndex}
        setSelectedIndex={setSelectedIndex}
        selectedIndexes={selectedIndexes}
        setSelectedIndexes={setSelectedIndexes}
        onCancel={onCancel}
        onConfirm={onConfirm as any}
        type={type as any}
      />
    </View>
  );
});

Picker.defaultProps = {
  editable: true,
  placeholder: '',
  type: 'single',
};

export interface PickerType extends BasePickerProps {
  selectedIndex: number;
  setSelectedIndex: (selectedIndex: number) => void;
  focus: () => void;
  focusIfEmpty: () => void;
}

export default withTailwind(Picker) as React.FC<
  PickerProps & { ref?: React.LegacyRef<PickerProps> }
>;
