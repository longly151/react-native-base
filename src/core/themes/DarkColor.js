module.exports = {
  dark: true,
  colors: {
    primary: '#219EE5',
    background: '#000000',
    card: '#1c1c1e',
    text: '#ffffff',
    border: '#272729',
    notification: '#ff453a',
    grey5: '#393e42',
    grey4: '#43484d',
    grey3: '#5e6977',
    grey2: '#86939e',
    grey1: '#bdc6cf',
    grey0: '#e1e8ee',
    white: '#080808',
    black: '#f2f2f2',
    info: '#36A3F7',
    success: '#34BFA3',
    warning: '#FFB822',
    error: '#F4516C',
  },
};
