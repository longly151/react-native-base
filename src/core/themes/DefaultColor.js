module.exports = {
  dark: false,
  colors: {
    primary: '#219EE5',
    background: '#f2f2f7',
    card: '#ffffff',
    text: '#000000',
    border: '#d8d8d8',
    notification: '#ff3b30',
    grey0: '#393e42',
    grey1: '#43484d',
    grey2: '#5e6977',
    grey3: '#86939e',
    grey4: '#bdc6cf',
    grey5: '#e1e8ee',
    white: '#ffffff',
    black: '#242424',
    info: '#36A3F7',
    success: '#34BFA3',
    warning: '#FFB822',
    error: '#F4516C',
  },
};
