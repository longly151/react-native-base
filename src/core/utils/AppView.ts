import { Font } from '@fonts';
import DefaultColor from '@themes/DefaultColor';
import { Dimensions, StyleProp, ViewStyle, TextStyle, ColorValue, ScaledSize } from 'react-native';
import ViewConfig from '@themes/ViewConfig';

const { width, height } = Dimensions.get('window');

/**
 * Shadow
 */
const androidDepth = {
  umbra: [
    '0px 2px 1px -1px',
    '0px 3px 1px -2px',
    '0px 3px 3px -2px',
    '0px 2px 4px -1px',
    '0px 3px 5px -1px',
    '0px 3px 5px -1px',
    '0px 4px 5px -2px',
    '0px 5px 5px -3px',
    '0px 5px 6px -3px',
    '0px 6px 6px -3px',
    '0px 6px 7px -4px',
    '0px 7px 8px -4px',
    '0px 7px 8px -4px',
    '0px 7px 9px -4px',
    '0px 8px 9px -5px',
    '0px 8px 10px -5px',
    '0px 8px 11px -5px',
    '0px 9px 11px -5px',
    '0px 9px 12px -6px',
    '0px 10px 13px -6px',
    '0px 10px 13px -6px',
    '0px 10px 14px -6px',
    '0px 11px 14px -7px',
    '0px 11px 15px -7px',
  ],
  penumbra: [
    '0px 1px 1px 0px',
    '0px 2px 2px 0px',
    '0px 3px 4px 0px',
    '0px 4px 5px 0px',
    '0px 5px 8px 0px',
    '0px 6px 10px 0px',
    '0px 7px 10px 1px',
    '0px 8px 10px 1px',
    '0px 9px 12px 1px',
    '0px 10px 14px 1px',
    '0px 11px 15px 1px',
    '0px 12px 17px 2px',
    '0px 13px 19px 2px',
    '0px 14px 21px 2px',
    '0px 15px 22px 2px',
    '0px 16px 24px 2px',
    '0px 17px 26px 2px',
    '0px 18px 28px 2px',
    '0px 19px 29px 2px',
    '0px 20px 31px 3px',
    '0px 21px 33px 3px',
    '0px 22px 35px 3px',
    '0px 23px 36px 3px',
    '0px 24px 38px 3px',
  ],
  ambient: [
    '0px 0px 0px 0px',
    '0px 1px 3px 0px',
    '0px 1px 5px 0px',
    '0px 1px 8px 0px',
    '0px 1px 10px 0px',
    '0px 1px 14px 0px',
    '0px 1px 18px 0px',
    '0px 2px 16px 1px',
    '0px 3px 14px 2px',
    '0px 3px 16px 2px',
    '0px 4px 18px 3px',
    '0px 4px 20px 3px',
    '0px 5px 22px 4px',
    '0px 5px 24px 4px',
    '0px 5px 26px 4px',
    '0px 6px 28px 5px',
    '0px 6px 30px 5px',
    '0px 6px 32px 5px',
    '0px 7px 34px 6px',
    '0px 7px 36px 6px',
    '0px 8px 38px 7px',
    '0px 8px 40px 7px',
    '0px 8px 42px 7px',
    '0px 9px 44px 8px',
    '0px 9px 46px 8px',
  ],
};
function interpolate(i: number, a: number, b: number, a2: number, b2: number) {
  return ((i - a) * (b2 - a2)) / (b - a) + a2;
}
function parseShadow(raw: any) {
  const values = raw.split(' ').map((val: any) => +val.replace('px', ''));
  return {
    x: values[0],
    y: values[1],
    blur: values[2],
    spread: values[3], // unused
  };
}
function onDepthChange(depth: number) {
  const s = parseShadow(androidDepth.penumbra[depth]);
  const y = s.y === 1 ? 1 : Math.floor(s.y * 0.5);
  return {
    elevation: depth + 1, // Android
    // iOS
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: y,
    },
    shadowOpacity: Number.parseFloat(interpolate(depth, 1, 24, 0.2, 0.6).toFixed(2)),
    shadowRadius: Number.parseFloat(interpolate(s.blur, 1, 38, 1, 16).toFixed(2)),
  };
}

class CAppView {
  private static _instance: CAppView;

  private constructor() {
    // ...
  }

  public static get Instance(): CAppView {
    /* istanbul ignore else */
    if (!this._instance) {
      this._instance = new this();
    }
    return CAppView._instance;
  }

  fontFamily: Font = 'GoogleSans-Regular';

  safeAreaInsets = {
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  };

  shadow = (depth: number = 10) => (depth >= 1 ? onDepthChange(depth - 1) : {});

  bottomNavigationBarHeight = 45;

  headerHeight = 50;

  bodyHeight =
    height -
    this.headerHeight -
    this.bottomNavigationBarHeight -
    this.safeAreaInsets.top -
    this.safeAreaInsets.bottom;

  bodyPaddingHorizontal = Number.parseFloat(ViewConfig.spacing.DEFAULT);

  bodyWidth = width - 2 * this.bodyPaddingHorizontal;

  roundedBorderRadius = Number.parseFloat(ViewConfig.borderRadius.DEFAULT);

  bottomSheetBorderRadius = Number.parseFloat(ViewConfig.borderRadius.card);

  itemMarginHorizontal = 10;

  screenWidth = width;

  screenHeight = height;

  isHorizontal = this.screenWidth > this.screenHeight;

  // Carousel
  sliderWidth = this.screenWidth;

  itemWidth = (this.isHorizontal ? 0.35 : 0.75) * this.screenWidth + this.itemMarginHorizontal * 2;

  itemHeight = 300;

  carouselBorderRadius = 8;

  onDimensionChange = (dimensions: { window: ScaledSize; screen: ScaledSize }) => {
    const { window } = dimensions;
    this.screenWidth = window.width;
    this.screenHeight = window.height;
    const isHorizontal = window.width > window.height;
    this.isHorizontal = isHorizontal;

    if (isHorizontal) {
      // Carousel
      this.itemWidth = 0.35 * window.width + this.itemMarginHorizontal * 2;
      this.sliderWidth = window.width;
    } else {
      this.itemWidth = 0.75 * window.width + this.itemMarginHorizontal * 2;
      this.sliderWidth = window.width;
    }
  };

  initSafeArea = (safeAreaInsets: { top: number; bottom: number; left: number; right: number }) => {
    const dimensions = Dimensions.get('window');
    this.safeAreaInsets = safeAreaInsets;
    this.bodyHeight =
      dimensions.height -
      this.headerHeight -
      this.bottomNavigationBarHeight -
      safeAreaInsets.top -
      safeAreaInsets.bottom;
  };

  getHeaderStyle = (theme: typeof DefaultColor) =>
    ({
      headerStyle: {
        backgroundColor: theme.colors.background,
      } as StyleProp<ViewStyle>,
      headerTitleStyle: {
        fontSize: 22,
        color: theme.colors.grey0,
        fontWeight: 'bold',
        fontFamily: AppView.fontFamily,
      } as StyleProp<TextStyle>,
      headerShadowVisible: false,
      headerBackTitleVisible: false,
      headerBackButtonMenuEnabled: false,
      headerTintColor: theme.colors.text,
    } as any);

  getQuickStyle = ({
    outline,
    clear,
    backgroundColor,
    textColor,
    borderColor,
  }: {
    outline?: boolean;
    clear?: boolean;
    backgroundColor: ColorValue;
    textColor: ColorValue;
    borderColor?: ColorValue;
  }) => {
    let quickStyles: StyleProp<ViewStyle> = {};
    let quickTextStyles: StyleProp<TextStyle> = {};
    if (outline) {
      quickStyles = {
        backgroundColor: 'transparent',
        borderColor: borderColor || backgroundColor,
        borderWidth: 1,
        elevation: 0,
      };
      quickTextStyles = {
        color: textColor || backgroundColor,
      };
    } else if (clear) {
      quickStyles = {
        backgroundColor: 'transparent',
        elevation: 0,
        borderRadius: 0,
      };
      quickTextStyles = { color: backgroundColor };
    } else {
      quickStyles = {
        backgroundColor,
      };
    }
    return { quickStyles, quickTextStyles };
  };

  getInputQuickStyle = ({
    outline,
    clear,
  }: {
    outline?: boolean;
    clear?: boolean;
    textColor: ColorValue;
  }) => {
    let quickStyles: StyleProp<ViewStyle> = {};
    if (outline) {
      quickStyles = {
        borderWidth: 1,
        elevation: 0,
      };
    } else if (clear) {
      quickStyles = {
        backgroundColor: 'transparent',
        borderRadius: 0,
        borderBottomWidth: 1,
        elevation: 0,
      };
    } else {
      quickStyles = {
        borderWidth: 1,
      };
    }
    return { quickStyles };
  };
}

const AppView = CAppView.Instance;
export default AppView;
