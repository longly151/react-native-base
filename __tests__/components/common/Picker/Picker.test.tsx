/* eslint-disable prefer-destructuring */
import React from 'react';
import { act, fireEvent, render, waitFor } from '@utils/TestHelper';
import Picker, { PickerType } from '@components/common/Picker/index';
import ExpoActionSheet, { ActionSheetProvider } from '@expo/react-native-action-sheet';
import DefaultColor from '@themes/DefaultColor';
import DarkColor from '@themes/DarkColor';
import AppView from '@utils/AppView';
import Color from 'color';
import { Platform } from 'react-native';
import * as Yup from 'yup';
import { Button } from '@components';
import { Formik } from 'formik';

global.console = {
  ...console,
  error: jest.fn(),
};

let mockTheme = DefaultColor;
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useTheme: jest.fn().mockImplementation(() => mockTheme),
}));

jest.mock('@expo/react-native-action-sheet', () => ({
  ...jest.requireActual('@expo/react-native-action-sheet'),
}));
const useActionSheet = jest.spyOn(ExpoActionSheet, 'useActionSheet');

const title = 'Select gender';
const options = ['Male', 'Female'];
const selectedIndex = 0;

beforeEach(() => {
  mockTheme = DefaultColor;
  Platform.OS = 'ios';
});

test('should render correctly', () => {
  const { toJSON, getByTestId } = render(<Picker options={['Male', 'Female']} />, {
    wrapper: ActionSheetProvider,
  });
  expect(toJSON()).toMatchSnapshot();
  expect(getByTestId('ContainerView')).toBeTruthy();
});

test('should have working props when unfocusing', () => {
  const { getByTestId, queryByTestId, rerender } = render(
    <Picker
      options={['Male', 'Female']}
      placeholderTextColor={mockTheme.colors.grey4}
      className="flex-1 flex-row self-center justify-between items-center rounded border mx-1 my-1 p-1 border-sky-700 bg-sky-500 text-white font-Roboto font-bold italic underline"
    />,
    {
      wrapper: ActionSheetProvider,
    },
  );

  // title, leftIcon, rightIcon should be null when the relevant props don't exist
  expect(queryByTestId('TitleText')).toBeFalsy();
  expect(queryByTestId('LeftIcon')).toBeFalsy();
  expect(queryByTestId('RightIcon')).toBeFalsy();

  // title, leftIcon, rightIcon should not be null when the relevant props exist
  const RootComponent = (
    <Picker
      title="Select gender"
      leftIconProps={{ name: 'search', size: 24 }}
      rightIconProps={{ name: 'close', size: 24 }}
      options={['Male', 'Female']}
      placeholderTextColor={mockTheme.colors.grey4}
      className="flex-1 flex-row self-center justify-between items-center rounded border mx-1 my-1 p-1 border-sky-700 bg-sky-500 text-white font-Roboto font-bold italic underline"
    />
  );
  rerender(RootComponent);

  expect(getByTestId('TitleText')).toBeTruthy();
  expect(getByTestId('LeftIcon')).toBeTruthy();
  expect(getByTestId('RightIcon')).toBeTruthy();

  // style testing
  expect(getByTestId('ContainerView')).toHaveStyle({
    flexGrow: 1,
    flexShrink: 1,
    flexDirection: 'row',
    alignSelf: 'center',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginLeft: 4,
    marginRight: 4,
    marginTop: 4,
    marginBottom: 4,
  });

  expect(getByTestId('PickerButton')).toHaveStyle({
    flexDirection: 'row',
    alignItems: 'center',
    borderLeftColor: 'rgba(3, 105, 161, 1)',
    borderRightColor: 'rgba(3, 105, 161, 1)',
    borderTopColor: 'rgba(3, 105, 161, 1)',
    borderBottomColor: 'rgba(3, 105, 161, 1)',
    backgroundColor: 'rgba(14, 165, 233, 1)',
    borderBottomLeftRadius: 4,
    borderBottomRightRadius: 4,
    borderTopLeftRadius: 4,
    borderTopRightRadius: 4,
    borderLeftWidth: 1,
    borderRightWidth: 1,
    borderTopWidth: 1,
    borderBottomWidth: 1,
    borderRadius: AppView.roundedBorderRadius,
    borderColor: Color(mockTheme.colors.grey5).darken(0.1).hex(), // unfocus borderColor
  });

  expect(getByTestId('OptionText')).toHaveStyle({
    paddingLeft: 4,
    paddingRight: 4,
    paddingTop: 4,
    paddingBottom: 4,
    color: mockTheme.colors.grey4, // placeholderColor
    fontFamily: 'Roboto',
    fontWeight: '700',
    fontStyle: 'italic',
    textDecorationLine: 'underline',
  });

  /**
   * DarkColor
   */
  mockTheme = DarkColor;
  const { getByTestId: getByTestIdDarkTheme } = render(RootComponent);

  expect(getByTestIdDarkTheme('PickerButton')).toHaveStyle({
    borderColor: mockTheme.colors.grey5, // unfocus borderColor
  });
});

test('should have working props when focusing', async () => {
  /**
   * Setup Test
   */
  const focusIfEmpty = jest.fn();
  const showActionSheetWithOptions = jest.fn();

  useActionSheet.mockReturnValue({
    showActionSheetWithOptions,
  });

  const RootComponent = (
    <Picker
      title={title}
      leftIconProps={{ name: 'search', size: 24 }}
      rightIconProps={{ name: 'close', size: 24 }}
      options={options}
      placeholderTextColor={mockTheme.colors.grey4}
      className="flex-1 flex-row self-center justify-between items-center rounded border mx-1 my-1 p-1 border-sky-700 bg-sky-500 text-white font-Roboto font-bold italic underline"
      nextFieldRef={{ current: { focusIfEmpty } }}
    />
  );
  const { getByTestId } = render(RootComponent, {
    wrapper: ActionSheetProvider,
  });

  /**
   * Press Button (Focus Trigger)
   */
  fireEvent.press(getByTestId('PickerButton'));

  const actionSheetConfig = showActionSheetWithOptions.mock.calls[0][0];
  const actionSheetCallback = showActionSheetWithOptions.mock.calls[0][1];

  expect(actionSheetConfig.title).toEqual(title);
  expect(actionSheetConfig.options).toEqual([...options, 'Cancel']);

  expect(getByTestId('PickerButton')).toHaveStyle({
    borderColor: mockTheme.colors.primary, // focus borderColor
  });

  expect(getByTestId('OptionText')).toHaveTextContent('');
  expect(getByTestId('OptionText')).toHaveStyle({
    color: mockTheme.colors.grey4, // placeholderColor
  });

  expect(showActionSheetWithOptions).toBeCalledTimes(1);

  /**
   * Select Item
   */
  await act(async () => {
    await actionSheetCallback(selectedIndex);
  });
  expect(getByTestId('OptionText')).toHaveTextContent(options[selectedIndex]);
  expect(getByTestId('OptionText')).toHaveStyle({
    color: 'rgba(255, 255, 255, 1)', // textColor
  });
  expect(actionSheetConfig.userInterfaceStyle).toEqual('light');

  await waitFor(() => expect(focusIfEmpty).toBeCalledTimes(1));

  /**
   * DarkColor
   */
  mockTheme = DarkColor;

  const showActionSheetWithOptionsDark = jest.fn();
  useActionSheet.mockReturnValue({
    showActionSheetWithOptions: showActionSheetWithOptionsDark,
  });
  const { getByTestId: getByTestIdDarkTheme } = render(RootComponent);

  fireEvent.press(getByTestIdDarkTheme('PickerButton'));
  const actionSheetConfigDark = showActionSheetWithOptionsDark.mock.calls[0][0];
  expect(actionSheetConfigDark.userInterfaceStyle).toEqual('dark');
});

test('should work correctly on iOS & Android', async () => {
  /**
   * iOS
   */
  Platform.OS = 'ios';
  const showActionSheetWithOptions = jest.fn();
  useActionSheet.mockReturnValue({
    showActionSheetWithOptions,
  });

  const { getByTestId, rerender } = render(<Picker title={title} options={options} />, {
    wrapper: ActionSheetProvider,
  });

  fireEvent.press(getByTestId('PickerButton'));

  const actionSheetCallback = showActionSheetWithOptions.mock.calls[0][1];
  const actionSheetConfig = showActionSheetWithOptions.mock.calls[0][0];

  expect(actionSheetConfig.destructiveButtonIndex).toEqual(undefined);

  await act(async () => {
    await actionSheetCallback(0);
  });

  /**
   * Android
   */
  Platform.OS = 'android';
  const showActionSheetWithOptionsAndroid = jest.fn();
  useActionSheet.mockReturnValue({
    showActionSheetWithOptions: showActionSheetWithOptionsAndroid,
  });

  rerender(<Picker title={title} options={options} />);
  fireEvent.press(getByTestId('PickerButton'));

  const actionSheetConfigAndroid = showActionSheetWithOptionsAndroid.mock.calls[0][0];
  expect(actionSheetConfigAndroid.destructiveButtonIndex).toEqual(options.length);

  await act(async () => {
    await actionSheetCallback(0);
  });
});

test('should work correctly with formik & nextField', async () => {
  /**
   * Setup Test
   */
  const showActionSheetWithOptions = jest.fn();
  useActionSheet.mockReturnValue({
    showActionSheetWithOptions,
  });

  const focusIfEmpty = jest.fn();
  const onCancel = jest.fn();

  const error = 'Gender is required';
  const formValidationSchema = Yup.object().shape({
    gender: Yup.number().required(error),
  });

  const onSubmit = jest.fn();

  const pickerRef: React.RefObject<PickerType> = React.createRef();
  const RootComponent = (
    <Formik
      validationSchema={formValidationSchema}
      initialValues={{
        gender: undefined,
      }}
      onSubmit={onSubmit}
    >
      {formikProps => (
        <>
          <Picker
            ref={pickerRef}
            formikProps={formikProps}
            name="gender"
            options={options}
            title={title}
            onCancel={onCancel}
            nextFieldRef={{ current: { focusIfEmpty } }}
          />
          <Button testID="SubmitButton" title="Submit" onPress={formikProps.handleSubmit} />
        </>
      )}
    </Formik>
  );
  const { getByTestId, queryByTestId, findByTestId } = render(RootComponent, {
    wrapper: ActionSheetProvider,
  });

  /**
   * Trigger focusIfEmpty
   */
  act(() => pickerRef.current?.focusIfEmpty());
  expect(getByTestId('PickerButton')).toHaveStyle({
    borderColor: mockTheme.colors.primary, // focus borderColor
  });

  /**
   * Trigger Error
   */
  fireEvent.press(getByTestId('SubmitButton'));

  expect(focusIfEmpty).toBeCalledTimes(0);
  expect(onSubmit).toBeCalledTimes(0);
  const errorText = await findByTestId('ErrorText');
  expect(errorText).toHaveTextContent(error);

  /**
   * Cancel ActionSheet
   */
  fireEvent.press(getByTestId('PickerButton'));

  const actionSheetCallback = showActionSheetWithOptions.mock.calls[0][1];
  await act(async () => {
    await actionSheetCallback(undefined);
  });
  expect(onCancel).toBeCalledTimes(1);

  /**
   * Successfully Select Item
   */
  fireEvent.press(getByTestId('PickerButton'));

  await act(async () => {
    await actionSheetCallback(selectedIndex);
  });

  expect(getByTestId('OptionText')).toHaveTextContent(options[selectedIndex]);

  fireEvent.press(getByTestId('SubmitButton'));
  await findByTestId('SubmitButton');

  await waitFor(() => expect(focusIfEmpty).toBeCalledTimes(1));
  expect(queryByTestId('ErrorText')).toBeFalsy();
  expect(onSubmit).toBeCalledTimes(1);
  expect(onSubmit).toHaveBeenCalledWith(
    {
      gender: selectedIndex,
    },
    expect.anything(),
  );
});
