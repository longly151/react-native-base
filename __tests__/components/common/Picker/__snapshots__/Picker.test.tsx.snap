// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={Object {}}
    testID="ContainerView"
  >
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#e1e8ee",
          "borderColor": "#C3D1DD",
          "borderRadius": 8,
          "flexDirection": "row",
          "opacity": 1,
        }
      }
      testID="PickerButton"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "GoogleSans-Regular",
            },
            Array [
              Object {},
              Array [
                Object {
                  "color": "#000000",
                  "flexDirection": "row",
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "fontSize": 16,
                  "overflow": "hidden",
                  "padding": 10,
                },
                Object {
                  "color": "#86939e",
                },
              ],
            ],
          ]
        }
        testID="OptionText"
      />
    </View>
  </View>
</View>
`;
